File: DP/121.cpp
/*
 * @Author: httermin
 * @Date: 2019-09-18 14:45:53
 */

#include <vector>
#include<iostream>
#include <algorithm>
#include<unordered_map>

using namespace std;


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(prices.size() <= 1) return 0;
        int maxV=0, minB;
        minB = prices[0]>prices[1]?prices[1]:prices[0];
        maxV = max(0, prices[1] - prices[0]);
        // cout<<maxV<<" "<<minB<<endl;

        for(int i=2; i<prices.size(); i++){
            maxV = max(maxV, prices[i] - minB);
            minB = min(minB, prices[i]);
            // cout<<maxV<<" "<<minB<<" "<<i<<endl;
        }
        return maxV;
    }
};

int main(){
    int A[] = {7, 3, 5, 1, 6, 4};
    vector<int> nums(A, A+6);
    Solution ss;
    cout<<ss.maxProfit(nums)<<endl;
}

File: DP/121.py
#-*-coding:utf-8-*-
'''
@Author: llei
@Date: 2022-06-05 20:13:06
'''
import sys
import os
from typing import List, Tuple, Dict

sys.path.append("..")

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) < 2:
            return 0

        max_prophit = 0
        min_price = prices[0]

        for i in range(1, len(prices)):
            max_prophit = max(max_prophit, prices[i] - min_price)
            min_price = min(min_price, prices[i])

        return max_prophit

if __name__ == "__main__":
    module = Solution()
    print(module.maxProfit([7,6,4,3,1]))